### PR TITLE
Add webhook sender

### PR DESCRIPTION
**Pull request by colm-prunty-cko**
This is a good PR!

### CODE REVIEW

File: WebhookSender.cs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+namespace SampleProject.Webhooks
+{
+    public class WebhookSender
+    {
+        private readonly HttpClient _httpClient;
+        private readonly IReportingService _reportingService;
+
+        public WebhookSender(HttpClient httpClient, IReportingService reportingService)
+        {
+            _httpClient = httpClient;
```
Comment (roman-pavlov-cko) [inline WebhookSender.cs:15]: Null check missed

File: WebhookSender.cs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+using System.Threading.Tasks;
+
+namespace SampleProject.Webhooks
+{
+    public class WebhookSender
+    {
+        private readonly HttpClient _httpClient;
+        private readonly IReportingService _reportingService;
+
+        public WebhookSender(HttpClient httpClient, IReportingService reportingService)
```
Comment (roman-pavlov-cko) [inline WebhookSender.cs:13]: Didn't assign the `reportingService ` variable

File: WebhookSender.cs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        {
+            _httpClient = httpClient;
+        }
+
+        public async Task SendWebhook(WebhookConfiguration configuration, bool webhookIsActive)
```
Comment (roman-pavlov-cko) [inline WebhookSender.cs:18]: paramater check

Comment (roman-pavlov-cko) [inline WebhookSender.cs:18]: `webhookIsActive` variable name can be better

File: WebhookSender.cs [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        {
+            var d = new WebhookDataStore("https://webhooks.checkout.com/datastore");
```
Comment (roman-pavlov-cko) [inline WebhookSender.cs:20]: coupling

Comment (roman-pavlov-cko) [inline WebhookSender.cs:20]: variable names

File: WebhookSender.cs [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+            var additionalWebhookInfo = d.GetAdditionalInfo(configuration.WebhookId);
+
+            var httpRequest = new HttpRequestMessage(HttpMethod.Post, configuration.Url);
+            var webhookResponse = _httpClient.Send(httpRequest);
```
Comment (roman-pavlov-cko) [inline WebhookSender.cs:24]: async?

File: WebhookSender.cs (hunk 3 above)
Comment (roman-pavlov-cko) [inline WebhookSender.cs:18]: Single responsibility of SendWebhook

File: WebhookSender.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+            {
+                await d.UpdateResult("Failed");
+            }
+
+            try
+            {
+                await _reportingService.UpdateReportingDatabase(configuration);
+            }
+            catch (Exception ex)
+            {
```
Comment (roman-pavlov-cko) [inline WebhookSender.cs:45]: handling logic?

File: WebhookSender.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+            else if (webhookResponse.StatusCode == HttpStatusCode.ServiceUnavailable)
+            {
+                await d.UpdateResult("Failed");
+            }
+
+            try
+            {
+                await _reportingService.UpdateReportingDatabase(configuration);
+            }
+            catch (Exception ex)
```
Comment (roman-pavlov-cko) [inline WebhookSender.cs:44]: Generic exception type,

File: WebhookSender.cs [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+                await d.UpdateResult("Failed");
+            }
+            else if (webhookResponse.StatusCode == HttpStatusCode.ServiceUnavailable)
+            {
+                await d.UpdateResult("Failed");
+            }
+
+            try
+            {
+                await _reportingService.UpdateReportingDatabase(configuration);
```
Comment (roman-pavlov-cko) [inline WebhookSender.cs:42]: Null

Comment (roman-pavlov-cko) [inline WebhookSender.cs:42]: What's the point?

File: WebhookSender.cs [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+            var d = new WebhookDataStore("https://webhooks.checkout.com/datastore");
+            var additionalWebhookInfo = d.GetAdditionalInfo(configuration.WebhookId);
+
+            var httpRequest = new HttpRequestMessage(HttpMethod.Post, configuration.Url);
+            var webhookResponse = _httpClient.Send(httpRequest);
+
+            if (webhookResponse.StatusCode == HttpStatusCode.OK)
+            {
+                await d.UpdateResult("Success");
+                configuration.Success = true;
```
Comment (roman-pavlov-cko) [inline WebhookSender.cs:29]: Why?

File: WebhookSender.cs (hunk 5 above)
Comment (roman-pavlov-cko) [inline WebhookSender.cs:24]: what if there is an exception in this line?
Retry?

File: WebhookSender.cs (hunk 3 above)
Comment (roman-pavlov-cko) [inline WebhookSender.cs:18]: Public docs?

File: WebhookSender.cs [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+            }
+            else if (webhookResponse.StatusCode == HttpStatusCode.NotFound)
+            {
+                await d.UpdateResult("Failed");
+            }
+            else if (webhookResponse.StatusCode == HttpStatusCode.ServiceUnavailable)
+            {
+                await d.UpdateResult("Failed");
+            }
```
Comment (roman-pavlov-cko) [inline WebhookSender.cs:38]: Long if-else.. - to a private method or out of this

File: WebhookSender.cs [hunk 11]
```diff
@@ -0,0 +1,49 @@
+using System;
+using System.Net.Http;
+using System.Net;
+using System.Threading.Tasks;
+
+namespace SampleProject.Webhooks
+{
+    public class WebhookSender
```
Comment (roman-pavlov-cko) [inline WebhookSender.cs:8]: Public comments?

File: WebhookSender.cs (hunk 2 above)
Comment (roman-pavlov-cko) [inline WebhookSender.cs:13]: Public comments?

File: WebhookSender.cs (hunk 3 above)
Comment (roman-pavlov-cko) [inline WebhookSender.cs:18]: `webhookIsActive ` not used

File: WebhookSender.cs (hunk 4 above)
Comment (roman-pavlov-cko) [inline WebhookSender.cs:20]: `d` variable name?

File: WebhookSender.cs [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+
+            var httpRequest = new HttpRequestMessage(HttpMethod.Post, configuration.Url);
+            var webhookResponse = _httpClient.Send(httpRequest);
+
+            if (webhookResponse.StatusCode == HttpStatusCode.OK)
+            {
+                await d.UpdateResult("Success");
+                configuration.Success = true;
+            }
+            else if (webhookResponse.StatusCode == HttpStatusCode.NotFound)
```
Comment (roman-pavlov-cko) [inline WebhookSender.cs:31]: Do we need to check all non-Ok code explicitly?

File: WebhookSender.cs [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+            {
+                await d.UpdateResult("Failed");
+            }
+            else if (webhookResponse.StatusCode == HttpStatusCode.ServiceUnavailable)
+            {
+                await d.UpdateResult("Failed");
```
Comment (roman-pavlov-cko) [inline WebhookSender.cs:37]: Should the name be `StoreResultAsync(result)`?

File: WebhookSender.cs [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+        {
+            _httpClient = httpClient;
+        }
+
+        public async Task SendWebhook(WebhookConfiguration configuration, bool webhookIsActive)
+        {
+            var d = new WebhookDataStore("https://webhooks.checkout.com/datastore");
+            var additionalWebhookInfo = d.GetAdditionalInfo(configuration.WebhookId);
+
+            var httpRequest = new HttpRequestMessage(HttpMethod.Post, configuration.Url);
```
Comment (roman-pavlov-cko) [inline WebhookSender.cs:23]: Do we send an empty body?

File: WebhookSender.cs (hunk 4 above)
Comment (Dgaduin) [inline WebhookSender.cs:20]: This might be better off as a dependency

File: WebhookSender.cs [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+
+        public WebhookSender(HttpClient httpClient, IReportingService reportingService)
+        {
+            _httpClient = httpClient;
+        }
+
+        public async Task SendWebhook(WebhookConfiguration configuration, bool webhookIsActive)
+        {
+            var d = new WebhookDataStore("https://webhooks.checkout.com/datastore");
+            var additionalWebhookInfo = d.GetAdditionalInfo(configuration.WebhookId);
```
Comment (Dgaduin) [inline WebhookSender.cs:21]: We are not using this variable at all

File: WebhookSender.cs (hunk 5 above)
Comment (Dgaduin) [inline WebhookSender.cs:24]: This request doesn't have any content, might be an error
Also might benefit of having a typed client for the calls - reduces dependencies on this level and can be mocked a lot easier for testing

File: WebhookSender.cs (hunk 1 above)
Comment (Dgaduin) [inline WebhookSender.cs:15]: We are not assigning the reporting service here

File: WebhookSender.cs (hunk 10 above)
Comment (Dgaduin) [inline WebhookSender.cs:38]: `switch` might be the better structure to use for these calls 
Also not all cases are considered

File: WebhookSender.cs (hunk 7 above)
Comment (Dgaduin) [inline WebhookSender.cs:44]: Catching a general exception which is also swallowed 
We should aim for specific exceptions which we might try and handle or at at least `throw;` the exception if we expect something up the chain to handle it.

File: WebhookSender.cs (hunk 8 above)
Comment (Dgaduin) [inline WebhookSender.cs:42]: It will depend on the code ideas but it might be a good idea to bundle the 2 remote actions into a transaction

**Pull request closed without merging**